### DISCUSSION TITLE
How to handle re-entrant functions?

### DISCUSSION THREAD

**Discussion by ptsouchlos** [danvratil/qcoro#123] [Q&A]
Following up from the discussion in #103, I'm experiencing an interesting problem while using a "heartbeat" with a `QSerialPort`. I'm note 100% sure if I'm using this term correctly, but it seems the `sendData()` function below is re-entrant. Based on the discussions in #103, we have the heartbeat coroutine polling the device but periodically we also need to send auxiliary commands to our device. However, when we do this, we interrupt the heartbeat before it finishes reading the response or we get the default command response instead of the true response when sending data directly using `sendData()`. 

Previously it was stated
> Regarding sending commands, it's safe to do so (as long as you are on the same thread as the other coroutines), since either the command() coroutine is running, or the heartbeat coroutines is running (or the event loop is just looping), there's no chance of a race. With serial port, I think it's also safe to assume that even if you send "Command A" immediatelly followed by "Command B" you will first receive "Response A" and then "Response B".

But I'm not sure this is actually the case unless I'm doing something wrong. Could how we're reading from the SerialPort be causing this issue?

Our basic setup is the following:

```cpp

QCoro::Task<> CommunicationDevice::heartbeat()
{
    QTimer timer;
    timer.start(static_cast<int>(std::round(1000.0 / update_frequency_.get())));
    while(device_->isOpen())
    {
        co_await timer;
        if(!device_->isOpen()) break;

        if(auto result = co_await sendData(default_command_))
        {
            default_command_processor_(result.value());
        }
        else
        {
            qCritical() << "Error occurred: " << result.error().reason;
        }
    }
}


QCoro::Task<Expected<QString, CommunicationError>> CommunicationDevice::sendData(const QString& data)
{

    long long bytes_written = -1;
    const auto send_buffer = data.toUtf8();
    device()->write(send_buffer);
    const auto& device_mode = device()->openMode();
    const auto& is_unbuffered = device_mode.testFlag(QIODevice::Unbuffered);
    if(const auto& bytes_written_opt = co_await qCoro(device()).waitForBytesWritten(-1))
        bytes_written = bytes_written_opt.value();

    // detect write error
    if(bytes_written != send_buffer.size() && !is_unbuffered)
    {
        qWarning() << "Bytes written mismatch! Wrote " << bytes_written << ", expected " << send_buffer.size();
        co_return tl::make_unexpected(CommunicationError{device()->errorString()});
    }

    // device is unbuffered so we expected the bytesWritten to be 0, check if we can read anything, if we can't then
    // that's actually an error
    if(is_unbuffered && device()->bytesAvailable() == 0)
    {
        qWarning() << "Unbuffered device write/read failed.";
        co_return tl::make_unexpected(CommunicationError{device()->errorString()});
    }

    // read from the device
    const auto reply = co_await read_strategy_(device());
    co_return reply;
}
```

We want to ensure that when we send the default command via the `heartbeat()` that we always get the default response and when we send auxiliary commands directly via `sendData()` we get the reponse for that command _only_. Any help is much appreciated.

**Comment by danvratil** [comment]
Hi Paul,

just to make sure I understand the problem correctly, what you are seeing is 

1) heartbeat calls `sendData()`
2) `sendData()` writes data to device and suspends while awaiting `read_strategy_()`, returning to event loop
3) `sendData()` is invoked from elsewhere to send an auxiliary command
4) the second invocation of `sendData()` writes data to device before reply to the first data has arrived

This is a problem inherent to any asynchronous code, even if you were not using coroutines, then you would most likely be `connected()`ed to the device's `readyRead()` signal, so after the first write, before the signal would be emitted, it would be possible for another write command to be issued.

In those scenarios I usually end up using a command queue - instead of directly invoking `sendData()`, I would enqueue the command (and receive a future or awaitable in return) into the command queue. If there's no ongoing write/read operation, the queue would invoke `sendData()` immediately, otherwise it would just append the command to a list of pending commands. When the currently ongoing operation receives response, it would complete the associated awaitable (so that the code that enqueued the particular command would receive the relevant response), and then check if there is another command in the queue to send. This way you can have multiple coroutines that are not aware of each other to request data to be sent, but you are guaranteed the data will be sent out in the right order and response is always received before next command is sent out.

If you are on Qt6 (sorry, I can't remember if you said you were using Qt5 or Qt6), you could just use `QPromise` inside the command queue and return an associated `QFuture` to the caller (QCoro has support for awaiting `QFuture`). If you are on Qt5, you can write a custom simple promise-awaitable pair.

**Reply by ptsouchlos** [reply]
Hi Dan, 

Yes that is correct and the command queue is the approach we were using previously before coroutines but I wasn't sure if there was something I was doing wrong now that we're exploring a coroutine approach. 

We're currently using Qt 5.15 LTS so I guess we'll have to write the await-able pair you mentioned.

**Reply by danvratil** [reply]
Let me know if you need any help with the awaitable :-)

**Reply by ptsouchlos** [reply]
A pointer in the right direction would be appreciated. 

My goal is to main the API that we have now from the "user's" perspective, namely:

```cpp
CommunicationDevice communication_device{...};
//...
communication_device.sendData(...).then([](const Expected<QString, CommunicationError> &result) {
	if(result) // do something with result
	else // do something with error
});
```

Based on our discussion though, this public facing `sendData()` function won't actually write data to the device, but will instead enqueue it to a command queue and return a promise-like awaitable. Does that seem reasonable?

**Reply by danvratil** [reply]
Sure, the caller doesn't really have to care whether `sendData()` actually sends the data or whether it just enqueues them and something else sends them later, as long as the returned awaitable eventually resolves with response to the command the caller has sent.

**Reply by ptsouchlos** [reply]
Ok great I think that's the route we will go then. 

When you mentioned a promise-awaiteable pair, did you mean something similar to `std::pair<>` or did you mean that we return an awaitable that also acts as a promise type?

**Reply by danvratil** [reply]
I meant the latter, or possibly even two coupled classes - a Promise-like and Future-like classes.

**Reply by ptsouchlos** [reply]
Sounds good.

With the command queue response, my plan is to use the heartbeat function to process commands in the queue. Any ideas on how to "wake" the heartbeat up to immediately prioritize new commands in the queue?

**Reply by ptsouchlos** [reply]
Here is what I came up with relatively quickly. I'm trying to leverage `QFuture` so I'm cheating by using the undocumented `QFutureInterface` class. 

`sendData()` now looks like this:

```cpp
QCoro::Task<Expected<QString, CommunicationError>> CommunicationDevice::sendData(const QString& data)
{
    auto future_interface = QSharedPointer<QFutureInterface<Expected<QString, CommunicationError>>>::create();
    command_queue_.emplace_back(std::make_pair(data, future_interface));
    // todo: wake up the heartbeat coroutine
    const auto& result = co_await qCoro(future_interface->future()).waitForFinished();
    co_return result;
}
```

And my heartbeat like this:

```cpp
QCoro::Task<> CommunicationDevice::heartbeat()
{
    QTimer timer;
    timer.start(static_cast<int>(std::round(1000.0 / update_frequency_.get())));
    while(device_->isOpen())
    {
        co_await timer;
        if(!device_->isOpen()) break;

        if(!command_queue_.empty())
        {
            // process commands
            for(auto& [command, future_interface] : command_queue_)
            {
                future_interface->reportStarted();
                const auto& cmd_result = co_await sendDataInternal(command);
                future_interface->reportResult(cmd_result);
                future_interface->reportFinished();
            }

            // clear all data
            command_queue_.clear();
        }
        // send the default command
        else
        {
            if(auto result = co_await sendDataInternal(default_command_))
            {
                default_command_processor_(result.value());
            }
            else
            {
                qCritical() << "Error occurred: " << result.error().reason;
            }
        }
    }
}
```

But the responses are still correct. It seems that I'm failing to ensure that the communication handshake completes before sending the next command. Any idea what I'm missing? I'm sure this is due to a gap in my knowledge of coroutines as I didn't have this problem when we were using threads 😅

**Reply by ptsouchlos** [reply]
The only way I've been able to get things work is to not use `co_await` in the `sendDataInternal()` function but this is not desirable as the GUI freezes.

**Reply by ptsouchlos** [reply]
Update:

I've been able to get a working solution. But I am now experiencing an odd crash when using qCoro with a signal.
Specifically, an exception is thrown on line 156 below. 
![image](https://user-images.githubusercontent.com/6591180/194162808-b12db07e-6483-4d67-983b-32094b09d2d5.png)

**Reply by danvratil** [reply]
Without seeing more, my guess would be that `this` is no longer valid when the slot is invoked - in the `connect()` statement you refer to `this->mObj` both as sender and receiver objects, so the connection won't get disconnected when the `this` object is destroyed, but you are still refering to `this` inside the slot.

Solution is to either use `this` as the receiver object in the `connect()` function (so the slot won't get called if the `this` object is destroyed before the signal is emitted), or make `this` object to be managed by a shared pointer (check `QEnableSharedFromThis`) so that it won't get destroyed until the lambda slot is disconnected.

**Reply by ptsouchlos** [reply]
Hi Dan, thank you for the feedback.

Here is what I have in the `heartbeat()`

```cpp
while(device_->isOpen())
{
    // check queue or send default command
    if(!command_queue_.empty())
    {
        while(!command_queue_.empty())
        {
            std::optional<CommandPair> opt_front = std::nullopt;
            try
            {
                std::lock_guard lock(command_queue_mutex_);
                opt_front = command_queue_.front();
                command_queue_.pop_front();
            }
            catch(...)
            {
            }
            if(opt_front)
            {
                auto [command, future_interface] = opt_front.value();
                future_interface->reportStarted();
                const auto& cmd_result = co_await writeDataToDevice(command);
                future_interface->reportResult(cmd_result);
                future_interface->reportFinished();
            }
        }
    }
    // send the default command
    else
    {
        if(auto result = co_await writeDataToDevice(default_command_))
        {
            default_command_processor_(result.value());
        }
        else
        {
            qCritical() << "Error occurred: " << result.error().reason;
        }
    }

    // check if the device is open
    if(!device_->isOpen()) break;

    // wait for signal to be emitted or the timeout
    co_await qCoro(&heartbeat_wakeup_, &detail::CommunicationDeviceHeartbeatNotifier::heartbeatWakeUp,
                   std::chrono::milliseconds(static_cast<std::size_t>(timeout)));
}
```

At the end, I'm trying to be clever and either wait for the heart beat wake signal to be emitted or we otherwise wait until our time has expired for the next round of sending the default command. 

And the heartbeat wake up object:
```cpp
class CommunicationDeviceHeartbeatNotifier final : public QObject
{
    Q_OBJECT;
public slots:
    void wakeUp() { emit heartbeatWakeUp(); }
Q_SIGNALS:
    void heartbeatWakeUp();
};
```

The error is happening when the following line is executed.
```cpp
 // wait for signal to be emitted or the timeout
co_await qCoro(&heartbeat_wakeup_, &detail::CommunicationDeviceHeartbeatNotifier::heartbeatWakeUp,
               std::chrono::milliseconds(static_cast<std::size_t>(timeout)));
```

Also I'm not sure I made it clear, but the snippet in my previous post is not my code but rather code in the qcoro library.

**Reply by danvratil** [reply]
Ah, I have not realized the screenshot in the previous comment from from QCoro code, sorry for the confusion :-)

I'll try to reduce it to some small reproducer case within QCoro and get a fix.

**Reply by danvratil** [reply]
Do you have any more details regarding the crash? What exception is thrown, possibly backtrace?

**Reply by ptsouchlos** [reply]
I'm not able to generate the crash at the moment as I have moved on to some other work, but there was an access violation being thrown in the qcoro signal helper [here](https://github.com/danvratil/qcoro/blob/main/qcoro/core/qcorosignal.h#L156).

**Reply by danvratil** [reply]
So far I was unable to reproduce the crash. I'll need more info to go on - at least a backtrace and if you are able to build and run your codebase under ASAN, then getting an ASAN report after the crash would help a lot.

**Reply by ptsouchlos** [reply]
I'll see what I can do. Thanks!
